File: BOJ/25_최단경로/9370_미확인도착지.cpp
#include <cstdio>
#include <iostream>
#include <queue>
#include <vector>
#include <algorithm>


#define MAX 2000
#define INF 987654321
typedef long long ll;
using namespace std;

int T,V,E,D,S,G,H;
vector<pair<int,int> > adjmat[MAX+1];
vector<int> dest;
ll dji(int st, int ed) {

    vector<int> dist(MAX+1,INF);
    priority_queue<pair<int,int> > pq;

    dist[st] = 0;
    pq.push(make_pair(0,st));
 
    while(!pq.empty()) {
        int cost = -pq.top().first;
        int here = pq.top().second;
        pq.pop();
        if(dist[here] < cost) continue;

        for(int i=0; i<adjmat[here].size(); i++) {
            int next = adjmat[here][i].first;
            int nextCost = cost + adjmat[here][i].second;

            if(dist[next] > nextCost) {
                dist[next] = nextCost;
                pq.push(make_pair(-nextCost,next));
            }
        }
    }
    return dist[ed];
}

int main() {
    
    scanf("%d",&T);
    for(int i=0; i<T; i++) {
        
        scanf("%d %d %d",&V,&E,&D);
        scanf("%d %d %d",&S,&G,&H);
        
        for(int j=0; j<MAX+1; j++) adjmat[j].clear();
        dest.clear();
        for(int j=0; j<E; j++) {
            int x,y,c;
            scanf("%d %d %d",&x,&y,&c);
            adjmat[x].push_back(make_pair(y,c));
            adjmat[y].push_back(make_pair(x,c));
        }
        
        for(int j=0; j<D; j++) {
            int d;
            scanf("%d",&d);
            dest.push_back(d);
        }


        vector<int> ans;
        for(int j=0; j<D; j++) {
            ll P1 = dji(S,G) + dji(G,H) + dji(H,dest[j]);
            ll P2 = dji(S,H) + dji(H,G) + dji(G,dest[j]);
            ll mins = dji(S,dest[j]);
            if (mins == min(P1,P2)) ans.push_back(dest[j]);
        }
   
        sort(ans.begin(),ans.end());
        for(int j=0; j<ans.size(); j++) printf("%d ",ans[j]);
        printf("\n");

    }

    return 0;
}  

File: BOJ/25_최단경로/9370_미확인도착지.py
import sys
import heapq

read = sys.stdin.readline
INF = sys.maxsize

def dji(st,ed) :

    dist = [INF for _ in range(N+1)]
    heap = []
    heapq.heappush(heap, (0,st))
    dist[st] = 0

    while heap :
        cost,cur = heapq.heappop(heap)
        if dist[cur] < cost : continue
        for nx,nxc in adjmat[cur] :
            nextCost = cost + nxc
            if dist[nx] > nextCost :
                dist[nx] = nextCost
                heapq.heappush(heap, (nextCost,nx))
    
    return dist[ed]

T = int(read())
for _ in range(T) :
    N,M,D = map(int,read().split())
    S,G,H = map(int,read().split())

    adjmat = [[] for _ in range(N+1)]
 
    for _ in range(M) :
        x,y,c = map(int, read().split())
        adjmat[x].append((y,c))
        adjmat[y].append((x,c))
    dest = [int(read()) for _ in range(D)]

    ans = []
    for d in dest :
        P1 = dji(S,G) + dji(G,H) + dji(H, d)
        P2 = dji(S,H) + dji(H,G) + dji(G, d)
        if dji(S,d) == min(P1,P2) :
            ans.append(d)
    ans = sorted(ans)
    for a in ans : print(a,end=" ")
    print()